### PR TITLE
[InstCombine] Add transforms for `(or/and (icmp eq/ne X,0),(icmp eq/ne X,Pow2OrZero))`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add tests for transforming `(or/and (icmp eq/ne X,0),(icmp eq/ne X,Pow2OrZero))`; NFC**
- **[InstCombine] Add transforms for `(or/and (icmp eq/ne X,0),(icmp eq/ne X,Pow2OrZero))`**

`(or (icmp eq X, 0), (icmp eq X, Pow2OrZero))`
    --> `(icmp eq (and X, Pow2OrZero), X)`

`(and (icmp ne X, 0), (icmp ne X, Pow2OrZero))`
    --> `(icmp ne (and X, Pow2OrZero), X)`

Proofs: https://alive2.llvm.org/ce/z/nPo2BN

### CODE REVIEW

**Comment by goldsteinn** [comment]
NB: this was approved by nikic a while back: https://reviews.llvm.org/D157312
There where concerns that we didn't handle `(icmp pred (x, y), x)` so well which has prompted a variety of independent patches to improve our handling of such cases.

Hopefully its time to get this in :)

**Comment by nikic** [comment]
Implementation still looks good to me. I think @dtcxzyw's tests should give us a much better answer about whether this canonicalization is problematic or not now :)

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 1]
```diff
@@ -701,6 +701,38 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
...
+                                    const SimplifyQuery &Q) {
+  CmpInst::Predicate Pred = IsAnd ? CmpInst::ICMP_NE : CmpInst::ICMP_EQ;
+  // Make sure we have right compares for our op.
+  if (LHS->getPredicate() != Pred || RHS->getPredicate() != Pred)
+    return nullptr;
+
+  // Make it so we can match LHS against the (icmp eq/ne X, 0) just for
+  // simplicity.
+  if (match(RHS->getOperand(1), m_Zero()))
+    std::swap(LHS, RHS);
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:720]: Missing test for this case.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 2]
```diff
@@ -701,6 +701,38 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
...
+
+  Value *Pow2, *Op;
+  // Match the desired pattern:
+  // LHS: (icmp eq/ne X, 0)
+  // RHS: (icmp eq/ne X, Pow2OrZero)
+  if (!match(LHS, m_ICmp(Pred, m_Value(Op), m_Zero())) ||
+      !match(RHS, m_c_ICmp(Pred, m_Specific(Op), m_Value(Pow2))) ||
+      !isKnownToBeAPowerOfTwo(Pow2, Q.DL, /*OrZero*/ true, /*Depth*/ 0, Q.AC,
+                              Q.CxtI, Q.DT))
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:729]:
```suggestion
      !isKnownToBeAPowerOfTwo(Pow2, Q.DL, /*OrZero=*/true, /*Depth=*/0, Q.AC,
                              Q.CxtI, Q.DT))
```

**Comment by dtcxzyw** [comment]
> 273 files changed, 45121 insertions(+), 45004 deletions(-)

We need a one-use check to avoid creating more instructions than before.

**Comment by nikic** [comment]
@dtcxzyw Could you please rerun tests with the new one-use checks?

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 3]
```diff
@@ -701,6 +701,42 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
...
+  // Match the desired pattern:
+  // LHS: (icmp eq/ne X, 0)
+  // RHS: (icmp eq/ne X, Pow2OrZero)
+  // We could potentially be more sophisticated than requiring LHS/RHS be
+  // one-use. We don't create additional instructions if only one of them is
+  // one-use. So cases where one is one-use and the other is two-use might be
+  // profitable.
+  if (!match(LHS, m_OneUse(m_ICmp(Pred, m_Value(Op), m_Zero()))) ||
+      !match(RHS, m_OneUse(m_c_ICmp(Pred, m_Specific(Op), m_Value(Pow2)))) ||
+      !isKnownToBeAPowerOfTwo(Pow2, Q.DL, /*OrZero=*/true, /*Depth=*/0, Q.AC,
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: We shouldn't fold `and (icmp ne X, 0), (icmp ne X, 1) -> icmp ne (and X, 1), X`.
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/667/files#r1632021232.

It will break the `x != 0 && x != 1 && x != 2` pattern:
https://github.com/php/php-src/blob/d1f14a46095571b1b6363fa7f83c6a681d1d02de/main/main.c#L453-L456

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: I guess it doesn't matter either way, it always turns into `x u> 1`: https://godbolt.org/z/4vE4nMKTb
although we should teach whichever pattern is looking at consecutative `ne` to handle that

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: Ill forbid `1`, also maybe should skip if `Op` is an `and` expression as well? That seemed to be the common element in all other regressions.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: Yeah, but there is a phase ordering problem blocking this fold:
https://github.com/dtcxzyw/llvm-opt-benchmark/pull/667/files#r1632021232

What do you think of the following fold?
https://alive2.llvm.org/ce/z/tmkPvm
```
define i1 @src(i16 %x) {
 %35 = and i16 %x, 255
 call void @use(i16 %35)
  %36 = and i16 %x, 1
  %or.cond = icmp ne i16 %36, %35
  ret i1 %or.cond
}

define i1 @tgt(i16 %x) {
  %35 = and i16 %x, 255
  call void @use(i16 %35)
  %cmp = icmp ugt i16 %35, 1
  ret i1 %cmp
}

declare void @use(i16)
```

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:733]: Fixed `1` cae + tests

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:733]: Emm, this regression still exists :( We may need more folds to fix this.

**Review by dtcxzyw — APPROVED** [review]
The old version https://github.com/llvm/llvm-project/commit/52823f5b0c5d329e1d34bec29526cfd4a3e0890f LGTM.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 4]
```diff
@@ -3240,6 +3277,11 @@ Value *InstCombinerImpl::foldAndOrOfICmps(ICmpInst *LHS, ICmpInst *RHS,
   ICmpInst::Predicate PredL = LHS->getPredicate(), PredR = RHS->getPredicate();
   Value *LHS0 = LHS->getOperand(0), *RHS0 = RHS->getOperand(0);
   Value *LHS1 = LHS->getOperand(1), *RHS1 = RHS->getOperand(1);
+  if (!IsLogical)
+    if (Value *V =
+            foldAndOrOfICmpsWithPow2AndWithZero(Builder, LHS, RHS, IsAnd, Q))
+      return V;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3283]: Why did you add this fold in this position? New exotic folds usually get added to the end, not the start of functions. Especially weird to see this interleaved with a bunch of variable declarations that you aren't using.

**Comment by nikic** [comment]
Did this fold produce *any* positive diffs? It seems like it's always either neutral (inverting a predicate) or a regression, but maybe I didn't see some.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 5]
```diff
@@ -701,6 +701,43 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
...
+  // Match the desired pattern:
+  // LHS: (icmp eq/ne X, 0)
+  // RHS: (icmp eq/ne X, Pow2OrZero)
+  // We could potentially be more sophisticated than requiring LHS/RHS be
+  // one-use. We don't create additional instructions if only one of them is
+  // one-use. So cases where one is one-use and the other is two-use might be
+  // profitable.
+  if (!match(LHS, m_OneUse(m_ICmp(Pred, m_Value(Op), m_Zero()))) ||
+      !match(RHS, m_OneUse(m_c_ICmp(Pred, m_Specific(Op), m_Value(Pow2)))) ||
+      !match(Pow2, m_One()) ||
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: > Did this fold produce _any_ positive diffs? It seems like it's always either neutral (inverting a predicate) or a regression, but maybe I didn't see some.

```suggestion
      match(Pow2, m_One()) ||
```

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: Now it only folds `and/or (icmp eq/ne X, 0), (icmp eq/ne X, 1)` :)

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: :O

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 4 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:3283]: Moving to the end.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 5 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Fixed :)

**Comment by dtcxzyw** [comment]
> Did this fold produce _any_ positive diffs? It seems like it's always either neutral (inverting a predicate) or a regression, but maybe I didn't see some.

See bench/php/optimized/php_date.ll

**Comment by goldsteinn** [comment]
ping

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 6]
```diff
@@ -701,6 +701,45 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
...
+  // Match the desired pattern:
+  // LHS: (icmp eq/ne X, 0)
+  // RHS: (icmp eq/ne X, Pow2OrZero)
+  // Skip if Pow2OrZero is 1. Either way it gets folded to (icmp ugt X, 1) but
+  // this form ends up slightly less canonical.
+  // We could potentially be more sophisticated than requiring LHS/RHS
+  // be one-use. We don't create additional instructions if only one
+  // of them is one-use. So cases where one is one-use and the other
+  // is two-use might be profitable.
+  if (!match(LHS, m_OneUse(m_ICmp(Pred, m_Value(Op), m_Zero()))) ||
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:732]: As a side note, we should probably add m_ICmp variants without predicate, as we have a bunch of places where the fetched predicate is just discarded, but kind of looks like it's part of the match. (Or maybe alternatively add something like m_SpecificICmp that matches the passed predicate instead of it being an out parameter.)

**Review by nikic — APPROVED** [review]
LGTM
